<!-- #######################################################################-->
<!-- -->
<!-- CCPN Data Model version 2.0.b3-->
<!-- -->
<!-- Autogenerated by PyApiDocGen revision 1.3 on Fri Sep 18 15:07:25 2009-->
<!--   from data model element ccp.molecule.Molecule revision 1.35.2.10-->
<!-- -->
<!-- #######################################################################-->

<!DOCTYPE HTML PUBLIC "-//W3C//DTD HTML 4.01 Transitional//EN"
"http://www.w3.org/TR/html4/loose.dtd">
<!--Generated by PyApiDocGen on Fri Sep 18 15:07:25 2009-->
<!--======================COPYRIGHT/LICENSE START==========================

index.html: Python API documentation for CCPN data model, package ccp.molecule.Molecule

Copyright (C) 2007  (CCPN Project)

=======================================================================

This library is free software; you can redistribute it and/or
modify it under the terms of the GNU Lesser General Public
License as published by the Free Software Foundation; either
version 2.1 of the License, or (at your option) any later version.
 
A copy of this license can be found in ../../../../../..license/LGPL.license
 
This library is distributed in the hope that it will be useful,
but WITHOUT ANY WARRANTY; without even the implied warranty of
MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the GNU
Lesser General Public License for more details.
 
You should have received a copy of the GNU Lesser General Public
License along with this library; if not, write to the Free Software
Foundation, Inc., 59 Temple Place, Suite 330, Boston, MA 02111-1307 USA


======================COPYRIGHT/LICENSE END============================

for further information, please contact :

- CCPN website (http://www.ccpn.ac.uk/)

- email: ccpn@bioc.cam.ac.uk

=======================================================================

If you are using this software for academic purposes, we suggest
quoting the following references:

===========================REFERENCE START=============================
Rasmus H. Fogh, Wayne Boucher, Wim F. Vranken, Anne
Pajon, Tim J. Stevens, T.N. Bhat, John Westbrook, John M.C. Ionides and
Ernest D. Laue (2005). A framework for scientific data modeling and automated
software development. Bioinformatics 21, 1678-1684.

===========================REFERENCE END===============================
-->
<HTML>
<head>
  <title>CCPN Python API - ccp.api.molecule.Molecule</title>
  <link rel=stylesheet type="text/css" href="../../../../../../doc/htmldoc.css">
  <script language="JavaScript">
    function wopn(href){ window.open(href,'','width=880,height=500,top=300,left=300,resizable=0,toolbar=0,location=0,directories=0,status=0,menubar=0,scrollbars=1')}
  </script>
</head>
<body>
<table bgcolor="#F0CC99" border="0" height="100" width="100%">
  <tr>
    <td>
      <a href="http://www.ccpn.ac.uk/" class="mainlogo"><img  alt="CCPN" border="0" id="toplink" src="../../../../../../doc/graphics/CCPN_Logo_200_x.gif"/></a>
      <br>
      &nbsp;&nbsp;&nbsp;
      Data Model version 2.0.b3
    </td>
    <td valign="center" width="80%">
      <!-- TOP BANNER BEGIN-->
      <div class="topbanner">
        <table border="0" width="100%">
          <tr valign="top">
            <td align="right" width="80%">
              <a href="javascript:wopn('../../../../../../doc/acknowledgements.html')">Acknowledgements</a>
            </td>
          </tr>
          <tr>
            <td align="center">
              <a href="../../../../../../model/doc/ccp/molecule/Molecule.html">Diagram</a>
              |
              <a href="../../../../../../python/doc/classMap.html">Class Map</a>
              |
              <a href="../../../../../../python/doc/attributeMap.html">Attribute Map</a>
              |
              <a href="../../../../../../python/doc/methodMap.html">Method Map</a>
              |
              <a href="javascript:wopn('../../../../../../doc/help.html')">Help</a>
              |
              <a href="javascript:wopn('../../../../../../doc/license.html')">License</a>
            </td>
          </tr>
          <tr valign="top">
            <td align="center" valign="bottom">
              <a href="../../../../../../python/doc/api.html">Home</a>
              |
              <span class="underover">
              Package:
              </span>
              {
              <a href="../../../../../../python/ccp/api/doc/molecule/MolSystem/index.html">prev</a>
              |
              <a href="../../../../../../python/ccp/api/doc/molecule/Stereochemistry/index.html">next</a>
              }
              |
              Class
              |
              Attribute
              |
              Method
            </td>
          </tr>
          <tr>
            <td>
            </td>
          </tr>
        </table>
      </div>
      <!-- TOP BANNER END-->
    </td>
  </tr>
</table>
<table cellpadding="5" width="100%">
  <tr>
    <td>
      <table cellpadding="2" cellspacing="2" width="100%">
        <tr valign="top">
          <td>
            <h2>Package: Molecule</h2>
            <ul>
            <h3>from Package <a href="../index.html">ccp.api.molecule</a></h3>
            </ul>
            <br>
            <ul>
            Covalent structure of molecules; sequence of linear poymers (DNA. RNA, and proteins).
            </ul>
          </td>
          <td align="right" valign="top" width="20%">
            <a href="#Classes">Classes</a>
            <br>
            <a href="#Data Types">Data Types</a>
            <br>
            <a href="#Data Obj Types">Data Obj Types</a>
          </td>
        </tr>
      </table>
      <br>
      <h3>Imported Packages:</h3>
      <a href="../../../../../../python/ccp/api/doc/molecule/ChemComp/index.html">ChemComp</a>, &nbsp;
      <a href="../../../../../../python/ccp/api/doc/molecule/ChemCompCoord/index.html">ChemCompCoord</a>, &nbsp;
      <a href="../../../../../../python/ccp/api/doc/general/Citation/index.html">ccp.api.general.Citation</a>, &nbsp;
      <a href="../../../../../../python/ccp/api/doc/general/DbRef/index.html">ccp.api.general.DbRef</a>, &nbsp;
      <a href="../../../../../../python/ccp/api/doc/general/Taxonomy/index.html">ccp.api.general.Taxonomy</a>, &nbsp;
      <a href="../../../../../../python/memops/api/doc/AccessControl/index.html">memops.api.AccessControl</a>, &nbsp;
      <a href="../../../../../../python/memops/api/doc/Implementation/index.html">memops.api.Implementation</a>
      <br>
      <br>
      <h3>Packages Importing ccp.api.molecule.Molecule:</h3>
      <a href="../../../../../../python/ccp/api/doc/molecule/LabeledMolecule/index.html">LabeledMolecule</a>, &nbsp;
      <a href="../../../../../../python/ccp/api/doc/molecule/MolSystem/index.html">MolSystem</a>, &nbsp;
      <a href="../../../../../../python/cambridge/api/doc/Dangle/index.html">cambridge.api.Dangle</a>, &nbsp;
      <a href="../../../../../../python/ccp/api/doc/general/Annotation/index.html">ccp.api.general.Annotation</a>, &nbsp;
      <a href="../../../../../../python/ccp/api/doc/lims/ExpBlueprint/index.html">ccp.api.lims.ExpBlueprint</a>, &nbsp;
      <a href="../../../../../../python/ccp/api/doc/lims/RefSampleComponent/index.html">ccp.api.lims.RefSampleComponent</a>, &nbsp;
      <a href="../../../../../../python/ccp/api/doc/nmr/Nmr/index.html">ccp.api.nmr.Nmr</a>, &nbsp;
      <a href="../../../../../../python/ccp/api/doc/nmr/NmrEntry/index.html">ccp.api.nmr.NmrEntry</a>, &nbsp;
      <a href="../../../../../../python/ccpnmr/api/doc/Analysis/index.html">ccpnmr.api.Analysis</a>, &nbsp;
      <a href="../../../../../../python/ccpnmr/api/doc/NameMapping/index.html">ccpnmr.api.NameMapping</a>
      <br>
    </td>
  </tr>
</table>
<table cellpadding="5" width="100%">
  <tr>
    <td>
      <br>
      <a name="Classes"></a>
      <h3>Classes:</h3>
      <table class="elemtable" cellpadding="2" cellspacing="2">
        <tr>
          <td class="tblhead">
            Class
          </td>
          <td class="tblhead">
            Description
          </td>
        </tr>
        <tr valign="top">
          <td class="tblrow">
            <a href="Alignment/index.html">Alignment</a>
          </td>
          <td class="tblrow">
            Alignment of molecule with database reference
          </td>
        </tr>
        <tr valign="top">
          <td class="tblrow">
            <a href="MolResLink/index.html">MolResLink</a>
          </td>
          <td class="tblrow">
            Link (bond) between two molResidues, through their MolResLinkEnds. MolResLinks that are valid for a Molecule are implicitly valid also for Chains linked to that Molecule. Unlike MolResLinkEnds, MolResLinks can be created (or modified) for existing molecules, even if the Molecule is linked to one or more Chains. This allows linkages to be redefined, or to be left as initially unknown (as in the case of a protein with unknown disulfide binding topology).
          </td>
        </tr>
        <tr valign="top">
          <td class="tblrow">
            <a href="MolResLinkEnd/index.html">MolResLinkEnd</a>
          </td>
          <td class="tblrow">
            The MolResidue end (Link End) of a bond to another MolResidue<br>
            The MolResLinkEnds corresponding to a MolResidue are created automatically when the MolResidue is created.<br>
            Note that MolResLinkEnds can neither be created nor deleted while their ancestor Molecule is linked to one or more chains.
          </td>
        </tr>
        <tr valign="top">
          <td class="tblrow">
            <a href="MolResidue/index.html">MolResidue</a>
          </td>
          <td class="tblrow">
            Monomer/Residue belonging to a Molecule, e.g. in the sequence for Insulin A-chain as such, as opposed to a specific Insulin A-chain molecule with coordinates.<br>
            <br>
            Note that MolResidues can neither be created nor deleted while their parent Molecule is linked to one or more chains. When creating a Molecule all relevant MolResLinkEnds are created automatically.<br>
            Note also that MolResidues can not be deleted, except as a consequence of deleting their molecule.<br>
            <br>
            It is guaranteed that the serials of molResidues form a continuous range, and that all linear polymer type links are between MolResidues with consecutive serials. The Molesidue.serials must start at one. This constraint is not checked (for technical reasons), but it can only be broken if someone has modified the data bypassing the API.
          </td>
        </tr>
        <tr valign="top">
          <td class="tblrow">
            <a href="MolSeqFragment/index.html">MolSeqFragment</a>
          </td>
          <td class="tblrow">
            Block of molResidues with consecutive serials in a molecule . For linear polymers this is guaranteed to be a connected fragment of the molecule. For other molecules the set of Molresidues selected may or may not be connected, depending on how the molecule was created. MolSeqFragments are disjoint and no residue can be a member of more than one.
          </td>
        </tr>
        <tr valign="top">
          <td class="tblrow">
            <a href="Molecule/index.html">Molecule</a>
          </td>
          <td class="tblrow">
            Abstract molecule, with defined topology but no location, coordinates, etc.
          </td>
        </tr>
        <tr valign="top">
          <td class="tblrow">
            <a href="MoleculeSysName/index.html">MoleculeSysName</a>
          </td>
          <td class="tblrow">
            Molecule systematic name in a given naming system.
          </td>
        </tr>
      </table>
    </td>
  </tr>
</table>
<table cellpadding="5" width="100%">
  <tr>
    <td>
      <br>
      <a name="Data Obj Types"></a>
      <h3>Data Obj Types:</h3>
      <table class="elemtable" cellpadding="2" cellspacing="2">
        <tr>
          <td class="tblhead">
            Type
          </td>
          <td class="tblhead">
            Description
          </td>
        </tr>
        <tr>
          <td class="tblrow" colspan="2">
            None
          </td>
        </tr>
      </table>
    </td>
  </tr>
</table>
<table cellpadding="5" width="100%">
  <tr>
    <td>
      <br>
      <a name="Data Types"></a>
      <h3>Data Types:</h3>
      <table class="elemtable" cellpadding="2" cellspacing="2">
        <tr>
          <td class="tblhead">
            Type
          </td>
          <td class="tblhead">
            Description
          </td>
        </tr>
        <tr valign="top">
          <td class="tblrow">
            <a href="LinkDihedralAngle.html">LinkDihedralAngle</a>
          </td>
          <td class="tblrow">
            Data Type for approximate dihedral angle across an interresidue bond, in degrees. Serves to specify the conformation across rotationally hindered bonds.
          </td>
        </tr>
        <tr valign="top">
          <td class="tblrow">
            <a href="MolType.html">MolType</a>
          </td>
          <td class="tblrow">
            This is an open enumeration for the type of Molecule and MolComponent.
          </td>
        </tr>
        <tr valign="top">
          <td class="tblrow">
            <a href="SecStrucCode.html">SecStrucCode</a>
          </td>
          <td class="tblrow">
            Secondary Structure Code, or DSSP code (Kabsch and Sander)<br>
            Values are :<br>
            H: 4-membered (alpha) helix<br>
            B: beta bridge<br>
            E: (extended) beta strand.<br>
            G: three-membered helix (3/01 helix)<br>
            I: five-membered (pi) helix<br>
            T: hydrogen-bonded turn<br>
            S: Bend (non-hydrogen bonded)<br>
            C: Random coil - not in a stable coformation<br>
            U: Undefined.
          </td>
        </tr>
        <tr valign="top">
          <td class="tblrow">
            <a href="SmilesType.html">SmilesType</a>
          </td>
          <td class="tblrow">
            Type of smiles string
          </td>
        </tr>
      </table>
    </td>
  </tr>
</table>

<table borderwidth="1" width="100%"><tr><td width="100%" bgcolor=#794E83></td></tr></table>
<table cellpadding="5" width="100%">
  <tr>
    <td>
      <table cellpadding="2" cellspacing="2">
        <tr valign="top">
          <td>
            <br><strong>Package guid:</strong>
          </td>
          <td>
            <br>www.ccpn.ac.uk_Fogh_2006-08-16-14:22:54_00038
          </td>
        </tr>
      </table>
    </td>
  </tr>
</table>


<table width="100%" cellpadding=2>
  <tr><td colspan=2 width="100%" bgcolor=#794E83></td></tr>
  <tr>
    <td><h5>
    
      &nbsp;
    <em>Data Model Version</em>&nbsp;2.0.b3
    </h5></td>
    <td align=right><a href='#toplink'><span style="font-size:9pt">Go&nbsp;to&nbsp;Top&nbsp;&nbsp;</span></a>
    </td>
  </tr>
</table>

<div class="leftbanner">
<table width=100% cellspacing=0 cellpadding=2 border=0>
  <tr><td bgcolor=#794E83>
    <nobr>
    <em>&nbsp;&nbsp;Autogenerated by</em>&nbsp;&nbsp;PyApiDocGen&nbsp;
    <em>revision</em>&nbsp;1.3&nbsp;&nbsp;
    <em>on</em>&nbsp;&nbsp;Fri Sep 18 15:07:25 2009&nbsp;</nobr>
    <nobr>&nbsp;
    <em>from data model package</em>&nbsp;&nbsp;ccp.api.molecule.Molecule&nbsp;&nbsp;
    <em>revision</em>&nbsp;1.35.2.10&nbsp;&nbsp;</nobr>
  </td></tr>
</table>
</div>
<table width=100% cellspacing=0 cellpadding=2 border=0>
  <tr valign=top>
    <td><h6>&nbsp;&nbsp;Work done by the <a href="mailto:ccpn-dev@mole.bio.cam.ac.uk">CCPN team</a>.</h6></td>
    <td align=right><h6><a target="blank" href="http://www.ccpn.ac.uk">www.ccpn.ac.uk</a>&nbsp;&nbsp;</h6></td>
  </tr>
</table>
<br>

</body>
</HTML>
